server/src/main/java/org/opensearch/index/engine/Engine.java

```diff
@@ -275,6 +281,35 @@ public void verifyEngineBeforeIndexClosing() throws IllegalStateException {
         }
     }
 
+    /**
+     * Get max sequence number that was part of last refresh.
+     * Sequence number is part of each document that is indexed.
+     * This method fetches the _id of last indexed document that was part of refresh and
+     * retrieves the _seq_no of the document.
+     */
+    public long getMaxSeqNoRefreshed(String source) throws IOException {
```

Comment:
I've been considering a similar change for segrep as well where we fetch the max seqNo contained in a set of segments, so that our seqNo passed to replicas is accurate to whats in the infos.  A few questions/thoughts.

1. Should we use a similar query to [restoreVersionmapAndCheckpointTracker](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L2858) that uses a range query with a starting seqNo?
2. IE is concurrently refreshing, meaning we could have a mismatch here depending on when the infos are fetched and this method is invoked.  What if we constructed a reader/IndexSearcher directly from the latest (to be uploaded) infos? Similar to NRTReplicationReaderManager:
3. 
```java
        try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
            SegmentInfos segmentInfos = segmentInfosGatedCloseable.get();
            DirectoryReader innerReader = StandardDirectoryReader.open(referenceToRefresh.directory(), segmentInfos, null, null);
            final IndexSearcher searcher = new IndexSearcher(innerReader);
            ...
        }
```
If we had a method that does this search based on the infos, we would reuse it with SR.

Reply:
> Should we use a similar query to [restoreVersionmapAndCheckpointTracker](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L2858) that uses a range query with a starting seqNo?

The query used in `restoreVersionmapAndCheckpointTracker` would be a bit expensive than the one we are using here. Any particular reason for using the same query?

> IE is concurrently refreshing, meaning we could have a mismatch here depending on when the infos are fetched and this method is invoked while using acquireSearcher on the engine.

Right. Currently, this method does not make any assumption on SegmentInfos and how it is used in conjunction with either SegRep or Remote Store. The method always return the maxSeqNo based on searcher's view of segments. I will add more info on the javadoc clarifying this.

But I got your point and is valid for remote store as well. The solution you provided would work.
I am thinking of overloading this method which takes `IndexSearcher` as input. What do you think?

Reply:
Would have to experiment but my thinking is the range query would be less expensive than the matchall, but I am no expert here in this optimization. 

Overloading and taking in an IndexSearcher or even a SegmentInfos would be great.

Reply:
+1 on 2nd point which Marc brought up. I think it makes sense to build methods that takes the segment infos that we are interested in. This way it is more deterministic.

Reply:
Requested review from @msfroh to understand the performance implications.

Reply:
Added method to accept `SegmentInfos`. We no longer require to fetch max seq number of last refresh (if required, can be added later).

Reply:
Changed the PR title accordingly.